codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/SyncClientClass.java

```diff
@@ -181,8 +200,37 @@ private MethodSpec constructor() {
         protocolSpec.errorResponseHandler(opModel).ifPresent(method::addCode);
 
         if (opModel.getEndpointDiscovery() != null) {
+            method.addStatement("boolean endpointDiscoveryEnabled = "
+                                + "clientConfiguration.option(SdkClientOption.ENDPOINT_DISCOVERY_ENABLED)");
+
+            if (opModel.getEndpointDiscovery().isRequired()) {
```

Comment:
This is complicated, but I'll have another go at the logic. 

The requirements state that for the values of the three flags (ED Required - ED enabled - Endpoint Overridden), 1. The general behavior when ED Required == TRUE is as follows:
If Endpoint Overridden == TRUE, there shall always be an exception
If ED enabled == FALSE there will also always be an exception
2. The customized flag allowEndpointOverrideForEndpointDiscoveryRequiredOperations should change the default behavior so that when Endpoint Overridden == TRUE instead of an exception we'll use the custom endpoint (by disabling the endpoint discovery...)

However, in the code below the Endpoint Overridden logic comes after ED enabled logic, so that if the values are FALSE - TRUE for ED enabled - Endpoint Overridden, we'll throw an exception even though it seems the requirement says we should use custom endpoints for all cases where it's specified (and the flag with the really long name is used). 

If I understood everything correct, either the requirement is incorrect, or we should switch places of these if statements and give endpoint overridden a higher priority.

Reply:
All of your assertions are correct except:

>However, in the code below the Endpoint Overridden logic comes after ED enabled logic, so that if the values are FALSE - TRUE for ED enabled - Endpoint Overridden, we'll throw an exception even though it seems the requirement says we should use custom endpoints for all cases where it's specified (and the flag with the really long name is used).

When ED enabled = false, we must ALWAYS throw an exception. The only case that allowEndpointOverrideForEndpointDiscoveryRequiredOperations should apply is the ED enabled = true case.

Reply:
Discussed offline, the spec changed a bit and we want to support what Anna-Karin asserted. Good catch.

Reply:
The refactored code is much easier to read, too!